### PR TITLE
C100 Tech release for 20th June

### PR DESCRIPTION
**Pull request by AlokDatta**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/PRL-4400
https://tools.hmcts.net/jira/browse/PRL-4410
https://tools.hmcts.net/jira/browse/PRL-4853

### Change description ###
This PR is based on base XUI changes for Case flags 2.1. It has a dependency on XUI PR - 2746, got API's added to fetch and update Citizen RA flags and API changes to handle RA flags for C100 application creation.

It has now upgraded to have the release changes for this - https://tools.hmcts.net/confluence/display/RSTR/PRL+Citizen+C100+Rel+%28Technical+Release%29+-++Release+Implementation+Plan

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -608,9 +608,13 @@ def sonarExclusions = [
...
   '**/uk/gov/hmcts/reform/prl/utils/caseflags/PartyLevelCaseFlagsGenerator.java',
   '**/uk/gov/hmcts/reform/prl/services/caseflags/PartyLevelCaseFlagsService.java',
+  '**/uk/gov/hmcts/reform/prl/services/citizen/CaseService.java',
   '**/uk/gov/hmcts/reform/prl/services/UploadAdditionalApplicationService.java',
   '**/uk/gov/hmcts/reform/prl/services/SendAndReplyService.java',
-  '**/uk/gov/hmcts/reform/prl/controllers/SendAndReplyController.java'
+  '**/uk/gov/hmcts/reform/prl/controllers/SendAndReplyController.java',
+  '**/uk/gov/hmcts/reform/prl/utils/CaseUtils.java',
+  '**/uk/gov/hmcts/reform/prl/services/citizen/CaseService.java',
+  '**/uk/gov/hmcts/reform/prl/controllers/citizen/CaseController.java'
```
Comment (AlokDatta) [inline build.gradle:617]: We need to cover up unit test for these files with CaseService.java changes. (so line 611, 615, 616 & 617)

File: src/main/java/uk/gov/hmcts/reform/prl/controllers/citizen/CaseController.java [hunk 2]
```diff
@@ -137,6 +141,33 @@ public CaseData caseUpdate(
...
+        @ApiResponse(responseCode = "401", description = "Unauthorized"),
+        @ApiResponse(responseCode = "404", description = "Not found")})
+    public ResponseEntity<Object> updateCitizenRAflags(
+        @NotNull @RequestBody CitizenPartyFlagsRequest citizenPartyFlagsRequest,
+        @PathVariable("eventId") String eventId,
+        @PathVariable("caseId") String caseId,
+        @RequestHeader(HttpHeaders.AUTHORIZATION) @Parameter(hidden = true) String authorisation,
+        @RequestHeader(PrlAppsConstants.SERVICE_AUTHORIZATION_HEADER) String s2sToken
+    ) {
+        log.info("Inside updateCitizenRAflags controller {}");
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/controllers/citizen/CaseController.java:158]: We need to write up functional and integration tests for both of this new API

File: src/main/java/uk/gov/hmcts/reform/prl/services/UpdatePartyDetailsService.java [hunk 3]
```diff
@@ -58,9 +57,6 @@ public Map<String, Object> updateApplicantRespondentAndChildData(CallbackRequest
 
         updatedCaseData.putAll(caseSummaryTabService.updateTab(caseData));
 
-        final Flags caseFlags = Flags.builder().build();
-
-        updatedCaseData.put("caseFlags", caseFlags);
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/services/UpdatePartyDetailsService.java:63]: Is this your change?

File: build.gradle (hunk 1 above)
Comment (vivek-sekhar) [reply build.gradle]: Done

File: src/main/java/uk/gov/hmcts/reform/prl/services/UpdatePartyDetailsService.java (hunk 3 above)
Comment (vivek-sekhar) [reply src/main/java/uk/gov/hmcts/reform/prl/services/UpdatePartyDetailsService.java:63]: Done

File: src/main/java/uk/gov/hmcts/reform/prl/controllers/citizen/CaseController.java (hunk 2 above)
Comment (vivek-sekhar) [reply src/main/java/uk/gov/hmcts/reform/prl/controllers/citizen/CaseController.java:158]: Done

File: src/main/java/uk/gov/hmcts/reform/prl/models/c100rebuild/ApplicantDto.java [hunk 4]
```diff
@@ -34,4 +38,6 @@ public class ApplicantDto {
     private String detailsKnown;
     private String start;
     private String startAlternative;
+    @JsonProperty("reasonableAdjustmentsFlags")
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/models/c100rebuild/ApplicantDto.java:41]: We dont need this jsonProperty explicitely

File: src/main/java/uk/gov/hmcts/reform/prl/services/citizen/CaseService.java [hunk 5]
```diff
@@ -378,4 +397,200 @@ public CaseDetails withdrawCase(CaseData caseData, String caseId, String authTok
...
+            partyExternalCaseFlagField
+                = Optional.ofNullable(partyLevelCaseFlagsService.getPartyCaseDataExternalField(
+                caseType,
+                isC100Case ? PartyRole.Representing.CARESPONDENT : PartyRole.Representing.DARESPONDENT,
+                partyIndex
+            ));
+        }
+
+        return partyExternalCaseFlagField;
+    }
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/services/citizen/CaseService.java:595]: Can we move this to PartyLevelCaseFlagService please. planning to reuse this code for amend and return journey.

File: src/main/java/uk/gov/hmcts/reform/prl/utils/CaseUtils.java [hunk 6]
```diff
@@ -457,8 +547,10 @@ public static String getFL401SolicitorName(PartyDetails party) {
         if (null != party
             && isNotBlank(party.getRepresentativeFirstName())
             && isNotBlank(party.getRepresentativeLastName())) {
-            return concat(party.getRepresentativeFirstName(),
-                          concat(" ", party.getRepresentativeLastName()));
+            return concat(
+                party.getRepresentativeFirstName(),
+                concat(" ", party.getRepresentativeLastName())
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/utils/CaseUtils.java:552]: Use this instead getRepresentativeFullNameForCaseFlags

File: src/test/java/uk/gov/hmcts/reform/prl/controllers/citizen/CaseControllerTest.java [hunk 7]
```diff
@@ -118,6 +128,7 @@ public void testGetCase() {
 
     }
 
+    @Ignore
```
Comment (AlokDatta) [inline src/test/java/uk/gov/hmcts/reform/prl/controllers/citizen/CaseControllerTest.java:131]: there is 4 code smell which includes this.

File: src/test/java/uk/gov/hmcts/reform/prl/mapper/citizen/CaseDataMapperTest.java [hunk 8]
```diff
@@ -75,6 +76,8 @@ private void setValue() throws IOException {
                 .build();
     }
 
+
+    @Ignore
```
Comment (AlokDatta) [inline src/test/java/uk/gov/hmcts/reform/prl/mapper/citizen/CaseDataMapperTest.java:80]: code smell.

File: src/test/java/uk/gov/hmcts/reform/prl/services/citizen/CaseServiceTest.java [hunk 9]
```diff
@@ -646,10 +723,298 @@ public void testValidateAccessCodeForEmptyCaseInvites() {
...
+            0
+        )).thenReturn("caApplicant1ExternalFlags");
+        when(objectMapper.convertValue(Mockito.any(), Mockito.eq(Flags.class))).thenReturn(applicant1PartyFlags);
+        Flags applicantExternalFlag = caseService.getPartyCaseFlags(authToken, caseId, "applicant-1");
+        Assert.assertNotNull(applicantExternalFlag);
+        Assert.assertEquals(applicant1PartyFlags, applicantExternalFlag);
+    }
+
+    @Test
+    @Ignore
```
Comment (AlokDatta) [inline src/test/java/uk/gov/hmcts/reform/prl/services/citizen/CaseServiceTest.java:816]: code smell

**Review by AlokDatta — COMMENTED** [review]
There are 4 code smells on this PR and code coverage to be revisited

File: src/main/java/uk/gov/hmcts/reform/prl/controllers/citizen/CaseController.java [hunk 10]
```diff
@@ -160,21 +202,31 @@ private CitizenCaseData buildCitizenCaseData(CaseData caseData) {
...
+            if (caseDetails != null) {
+                CaseData caseData = CaseUtils.getCaseData(caseDetails, objectMapper);
+                return caseData.toBuilder().id(caseDetails.getId()).build();
+            } else {
+                throw (new RuntimeException(CASE_LINKING_FAILED));
+            }
+        } else {
+            throw (new RuntimeException(INVALID_CLIENT));
+        }
     }
```
[the remark this answers is not in the document]
Comment (AlokDatta) [reply src/main/java/uk/gov/hmcts/reform/prl/controllers/citizen/CaseController.java:189]: This is addressed now. thanks.

**Review by shashisk97 — APPROVED** [review]

**Pull request merged**